zerver/views/invite.py

```diff
@@ -134,6 +134,30 @@ def invite_users_backend(
     group_ids: Json[list[int]] | None = None,
     include_realm_default_subscriptions: Json[bool] = False,
 ) -> HttpResponse:
+    # Moved permissions checks to a helper function to not violate SOLID principles
+    validate_invite_permissions(user_profile, invite_as)
+    # Moved email stuff to help with readability and maintainability (and SOLID)
+    invitee_emails = parse_invitee_emails(invitee_emails_raw)
+    streams = access_streams_for_invite(stream_ids, user_profile)
+    user_groups = access_user_groups_for_invite(group_ids, user_profile)
+
+    # Moved invite processing into a function to fix SOLID violations
+    skipped = process_invite(
+        user_profile,
+        invitee_emails,
+        streams,
+        notify_referrer_on_join,
+        user_groups,
+        invite_expires_in_minutes,
+        include_realm_default_subscriptions,
+        invite_as,
+    )
+
+    # Invite response handling moved to function
+    return handle_invite_response(request, skipped)
+
+
+def validate_invite_permissions(user_profile: UserProfile, invite_as: int) -> None:
```

Comment:
This is almost an ok abstraction, but I'd prefer a clean commit that just creates this function, placing it above `invite_users_backend`, and it probably a name like `check_invite_user_permission(user_profile, invite_as)`. The problem is it does a bad job of interacting with `generate_multiuse_invite_backend`. You'll need to think about this more.